### PR TITLE
Detect cycles in object graph to be serialized.

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: genson/src/main/java/com/owlike/genson/convert/RuntimeTypeConverter.java [hunk 1]
```diff
@@ -58,4 +69,32 @@ private boolean isContainer(T obj) {
...
     }
+
+    private boolean isSimpleType(T obj) {
+      return obj instanceof Boolean ||
+          obj instanceof Number ||
+          obj instanceof String;
+    }
+
+    private void ensureNoCircularRefs(T obj, Context ctx) {
+      if (!isContainer(obj) && !isSimpleType(obj))
```
Comment (aseovic) [inline genson/src/main/java/com/owlike/genson/convert/RuntimeTypeConverter.java:80]: Do you need this check? Seems redundant, as this method is only called if the condition above is true.

File: genson/src/main/java/com/owlike/genson/convert/RuntimeTypeConverter.java [hunk 2]
```diff
@@ -58,4 +69,32 @@ private boolean isContainer(T obj) {
...
+      {
+        Set<Object> seen = ctx.get(CYCLE_KEY, Set.class);
```
Comment (aseovic) [inline genson/src/main/java/com/owlike/genson/convert/RuntimeTypeConverter.java:82]: Should we use `IdentityHashMap` here? It is probably more efficient than a `HashSet`, and is pretty much designed for this particular purpose.

**Review by aseovic — APPROVED** [review]

File: genson/src/main/java/com/owlike/genson/convert/RuntimeTypeConverter.java (hunk 1 above)
Comment (ghost) [reply genson/src/main/java/com/owlike/genson/convert/RuntimeTypeConverter.java:80]: Yep, it's intentional.  It's just an added check to ensure we're not tracking anything that we don't need to track.  There won't be cycles in Strings, Boolean, Numbers.  Maybe overkill.

File: genson/src/main/java/com/owlike/genson/convert/RuntimeTypeConverter.java (hunk 2 above)
Comment (ghost) [reply genson/src/main/java/com/owlike/genson/convert/RuntimeTypeConverter.java:82]: Good point.

**Pull request merged**